dev/bots/test.dart

```diff
@@ -241,6 +243,132 @@ Future<Null> _runFlutterAnalyze(String workingDirectory, {
   );
 }
 
+Future<Null> _verifyNoBadImports(String workingDirectory) async {
+  final List<String> errors = <String>[];
+  final String libPath = path.join(workingDirectory, 'packages', 'flutter', 'lib');
+  final String srcPath = path.join(workingDirectory, 'packages', 'flutter', 'lib', 'src');
+  // Verify there's one libPath/*.dart for each srcPath/*/.
+  <String>[];
+  final List<String> packages = new Directory(libPath).listSync()
+    .where((FileSystemEntity entity) => entity is File && path.extension(entity.path) == '.dart')
+    .map<String>((FileSystemEntity entity) => path.basenameWithoutExtension(entity.path))
+    .toList()..sort();
+  final List<String> directories = new Directory(srcPath).listSync()
+    .where((FileSystemEntity entity) => entity is Directory)
+    .map<String>((FileSystemEntity entity) => path.basename(entity.path))
+    .toList()..sort();
+  if (!_matches(packages, directories)) {
+    errors.add(
+      'flutter/lib/*.dart does not match flutter/lib/src/*/:\n'
+      'These are the exported packages:\n' +
+      packages.map((String path) => '  lib/$path.dart').join('\n') +
+      'These are the directories:\n' +
+      directories.map((String path) => '  lib/src/$path/').join('\n')
+    );
+  }
+  // Verify that the imports are well-ordered.
+  final Map<String, Set<String>> dependencyMap = new Map<String, Set<String>>.fromIterable(
+    directories,
+    key: (String directory) => directory,
+    value: (String directory) => _findDependencies(path.join(srcPath, directory), errors, checkForMeta: directory != 'foundation'),
+  );
+  for (String package in dependencyMap.keys) {
+    if (dependencyMap[package].contains(package)) {
+      errors.add(
+        'One of the files in the $yellow$package$reset package imports that package recursively.'
+      );
+    }
+  }
+  for (String package in dependencyMap.keys) {
+    final List<String> loop = _deepSearch(dependencyMap, package);
+    if (loop != null) {
+      errors.add(
+        '${yellow}Dependency loop:$reset ' +
+        loop.join(' depends on ')
+      );
+    }
+  }
+  // Fail if any errors
+  if (errors.isNotEmpty) {
+    print('$red━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━$reset');
+    if (errors.length == 1) {
+      print('${bold}An error was detected when looking at import dependencies within the Flutter package:$reset\n');
+    } else {
+      print('${bold}Multiple errors were detected when looking at import dependencies within the Flutter package:$reset\n');
+    }
+    print(errors.join('\n\n'));
+    print('$red━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━$reset\n');
+    exit(1);
+  }
+}
+
+bool _matches<T>(List<T> a, List<T> b) {
+  assert(a != null);
```

Comment:
How strongly do you feel about this? I think the code I have now is both more efficient and more readable, but I don't feel strongly about this.